halo2_proofs/src/plonk/circuit.rs

```diff
@@ -599,6 +600,173 @@ impl<F: Field> Expression<F> {
+                    sum,
+                    product,
+                    scaled,
+                    zero,
+                );
+                negated(a)
+            }
+            Expression::Sum(a, b) => {
+                let a = a.evaluate_lazy(
+                    constant,
+                    selector_column,
+                    fixed_column,
+                    advice_column,
+                    instance_column,
+                    negated,
+                    sum,
+                    product,
+                    scaled,
+                    zero,
+                );
+                let b = b.evaluate_lazy(
+                    constant,
+                    selector_column,
+                    fixed_column,
+                    advice_column,
+                    instance_column,
+                    negated,
+                    sum,
+                    product,
+                    scaled,
+                    zero,
+                );
+                sum(a, b)
+            }
+            Expression::Product(a, b) => {
+                let (a, b) = if a.complexity() <= b.complexity() {
+                    (a, b)
+                } else {
+                    (b, a)
+                };
+                let a = a.evaluate_lazy(
+                    constant,
+                    selector_column,
+                    fixed_column,
+                    advice_column,
+                    instance_column,
+                    negated,
+                    sum,
+                    product,
+                    scaled,
+                    zero,
+                );
+
+                if a == *zero {
+                    a
+                } else {
+                    let b = b.evaluate_lazy(
+                        constant,
+                        selector_column,
+                        fixed_column,
+                        advice_column,
+                        instance_column,
+                        negated,
+                        sum,
+                        product,
+                        scaled,
+                        zero,
+                    );
+                    product(a, b)
+                }
+            }
+            Expression::Scaled(a, f) => {
+                let a = a.evaluate_lazy(
+                    constant,
+                    selector_column,
+                    fixed_column,
+                    advice_column,
+                    instance_column,
+                    negated,
+                    sum,
+                    product,
+                    scaled,
+                    zero,
+                );
+                scaled(a, *f)
+            }
+        }
+    }
+
+    /// Identifier for this expression. Expressions with identical identifiers
+    /// do the same calculation (but the expressions don't need to be exactly equal
+    /// in how they are composed e.g. `1 + 2` and `2 + 1` can have the same identifier).
+    pub fn identifier(&self) -> String {
```

Comment:
> ... but the expressions don't need to be exactly equal in how they are composed e.g. `1 + 2` and `2 + 1` can have the same identifier

At first glance at the docs I thought this method helps to organize expression to make `1 + 2` and `2 + 1` have the same identifier, but it seems not. Does this aim to info developer to use more identical expression as possible (to reuse stored expression as much as possible)?

Reply:
I had some naive code to try and do the reorganization, but chose to remove it for now because it didn't really work well enough and didn't make any difference. I think it's not super easy to do this well elegantly (or at least I haven't found one yet, but haven't really tried much to be honest). But yeah the goal is indeed that this identifier can directly be used to detect expressions that are interchangeable (so in our use case to reuse stored expressions when possible).

So for now the docs share how the function could behave in the future, and so e.g. should not be used to compare expressions if strict equality is required.

Reply:
I see, I also tried once to normalize expression, but I found it's really ugly (use some kinds of map as key of map) and not sure if it works, can feel that it's not easy to do.

Reply:
Since `HashMap` in std rust now relies on `hashbrown` which is pretty damn fast. 
Have we tried to store as ID the map key containing `([values, involved], Expression::Type)`? With arrays we can check quickly with `contains` and similar things. Same in Iterators.

`Expression::Type` here simply means that for any set of inputs (independent of the order), if we perform the same op we will get the same result. (So we don't need to store the result in the keys and only in the values).

Is there anything painful on using HashMap?